diff --git a/dist/src/listener.js b/dist/src/listener.js
index 8b0edd335519bbcd6911473f0418bdbaf803f32d..cc527c2d5267d279a475bebf3cf7cc8f0ad74c24 100644
--- a/dist/src/listener.js
+++ b/dist/src/listener.js
@@ -140,27 +140,21 @@ export class TCPListener extends TypedEventEmitter {
         }
         this.log('new inbound connection %s', maConn.remoteAddr);
         this.sockets.add(socket);
+        socket.once('close', () => {
+            this.sockets.delete(socket);
+            if (this.context.closeServerOnMaxConnections != null &&
+                this.sockets.size < this.context.closeServerOnMaxConnections.listenBelow) {
+                this.resume().catch(err => {
+                    this.log.error('error attempting to listen server once connection count under limit - %e', err);
+                    this.context.closeServerOnMaxConnections?.onListenError?.(err);
+                });
+            }
+        });
         this.context.upgrader.upgradeInbound(maConn, {
             signal: this.shutdownController.signal
         })
             .then(() => {
             this.log('inbound connection upgraded %s', maConn.remoteAddr);
-            socket.once('close', () => {
-                this.sockets.delete(socket);
-                if (this.context.closeServerOnMaxConnections != null &&
-                    this.sockets.size < this.context.closeServerOnMaxConnections.listenBelow) {
-                    // The most likely case of error is if the port taken by this
-                    // application is bound by another process during the time the
-                    // server if closed. In that case there's not much we can do.
-                    // resume() will be called again every time a connection is
-                    // dropped, which acts as an eventual retry mechanism.
-                    // onListenError allows the consumer act on this.
-                    this.resume().catch(err => {
-                        this.log.error('error attempting to listen server once connection count under limit - %e', err);
-                        this.context.closeServerOnMaxConnections?.onListenError?.(err);
-                    });
-                }
-            });
             if (this.context.closeServerOnMaxConnections != null &&
                 this.sockets.size >= this.context.closeServerOnMaxConnections.closeAbove) {
                 this.log('pausing incoming connections as limit is exceeded - %d/%d', this.sockets.size, this.context.closeServerOnMaxConnections.closeAbove);
@@ -170,7 +164,6 @@ export class TCPListener extends TypedEventEmitter {
             .catch(async (err) => {
             this.log.error('inbound connection upgrade failed - %e', err);
             this.metrics.errors?.increment({ [`${this.addr} inbound_upgrade`]: true });
-            this.sockets.delete(socket);
             maConn.abort(err);
         });
     }
@@ -220,10 +213,14 @@ export class TCPListener extends TypedEventEmitter {
         // synchronously close any open connections - should be done after closing
         // the server socket in case new sockets are opened during the shutdown
         this.sockets.forEach(socket => {
-            if (socket.readable) {
-                events.push(pEvent(socket, 'close', options));
-                socket.destroy();
+            if (socket.closed) {
+                return;
             }
+            events.push(pEvent(socket, 'close', {
+                ...options,
+                rejectionEvents: []
+            }));
+            socket.destroy();
         });
         await Promise.all(events);
     }
diff --git a/dist/src/tcp.js b/dist/src/tcp.js
index 7e59cb40a94721669d392872c0ca431dec10b162..dea9fa3e4f44491ca63715c4e2828fb77c94c885 100644
--- a/dist/src/tcp.js
+++ b/dist/src/tcp.js
@@ -27,8 +27,9 @@
  * ```
  */
 import net from 'net';
-import { AbortError, TimeoutError, serviceCapabilities, transportSymbol } from '@libp2p/interface';
+import { AbortError, NotStartedError, TimeoutError, serviceCapabilities, transportSymbol } from '@libp2p/interface';
 import { TCP as TCPMatcher } from '@multiformats/multiaddr-matcher';
+import { setMaxListeners } from 'main-event';
 import { CustomProgressEvent } from 'progress-events';
 import { TCPListener } from './listener.js';
 import { toMultiaddrConnection } from './socket-to-conn.js';
@@ -38,10 +39,15 @@ export class TCP {
     metrics;
     components;
     log;
+    sockets;
+    shutdownController;
     constructor(components, options = {}) {
         this.log = components.logger.forComponent('libp2p:tcp');
         this.opts = options;
         this.components = components;
+        this.sockets = new Set();
+        this.shutdownController = new AbortController();
+        setMaxListeners(Infinity, this.shutdownController.signal);
         if (components.metrics != null) {
             this.metrics = {
                 events: components.metrics.registerCounterGroup('libp2p_tcp_dialer_events_total', {
@@ -60,7 +66,20 @@ export class TCP {
     [serviceCapabilities] = [
         '@libp2p/transport'
     ];
+    start() {
+        this.shutdownController = new AbortController();
+        setMaxListeners(Infinity, this.shutdownController.signal);
+    }
+    async stop() {
+        this.shutdownController.abort();
+        await Promise.all([...this.sockets].map(async (socket) => {
+            await this.closeSocket(socket);
+        }));
+    }
     async dial(ma, options) {
+        if (this.shutdownController.signal.aborted) {
+            throw new NotStartedError('TCP transport is stopped');
+        }
         options.keepAlive = options.keepAlive ?? true;
         options.noDelay = options.noDelay ?? true;
         options.allowHalfOpen = options.allowHalfOpen ?? false;
@@ -105,6 +124,10 @@ export class TCP {
             });
             this.log('dialing %a with opts %o', ma, cOpts);
             rawSocket = net.connect(cOpts);
+            this.sockets.add(rawSocket);
+            rawSocket.once('close', () => {
+                this.sockets.delete(rawSocket);
+            });
             const onError = (err) => {
                 this.log.error('dial to %a errored - %e', ma, err);
                 const cOptsStr = cOpts.path ?? `${cOpts.host ?? ''}:${cOpts.port}`;
@@ -136,6 +159,7 @@ export class TCP {
                 if (options.signal != null) {
                     options.signal.removeEventListener('abort', onAbort);
                 }
+                this.shutdownController.signal.removeEventListener('abort', onAbort);
                 if (err != null) {
                     reject(err);
                     return;
@@ -146,12 +170,27 @@ export class TCP {
             rawSocket.on('timeout', onTimeout);
             rawSocket.on('connect', onConnect);
             options.signal.addEventListener('abort', onAbort);
+            this.shutdownController.signal.addEventListener('abort', onAbort);
         })
-            .catch(err => {
-            rawSocket?.destroy();
+            .catch(async (err) => {
+            if (rawSocket != null) {
+                await this.closeSocket(rawSocket);
+            }
             throw err;
         });
     }
+    async closeSocket(socket) {
+        if (socket.closed) {
+            return;
+        }
+        const closed = new Promise(resolve => {
+            socket.once('close', () => {
+                resolve();
+            });
+        });
+        socket.destroy();
+        await closed;
+    }
     /**
      * Creates a TCP listener. The provided `handler` function will be called
      * anytime a new incoming Connection has been successfully upgraded via
diff --git a/src/listener.ts b/src/listener.ts
index 7217f0423f793eaad20374da7e50f248f8429eb0..45ceb98bf3977f78d8236265be56d1b7337641cc 100644
--- a/src/listener.ts
+++ b/src/listener.ts
@@ -179,33 +179,22 @@ export class TCPListener extends TypedEventEmitter<ListenerEvents> implements Li
 
     this.log('new inbound connection %s', maConn.remoteAddr)
     this.sockets.add(socket)
-
+    socket.once('close', () => {
+         this.sockets.delete(socket);
+         if (this.context.closeServerOnMaxConnections != null &&
+             this.sockets.size < this.context.closeServerOnMaxConnections.listenBelow) {
+             this.resume().catch(err => {
+                 this.log.error('error attempting to listen server once connection count under limit - %e', err);
+                 this.context.closeServerOnMaxConnections?.onListenError?.(err);
+             });
+         }
+     });
     this.context.upgrader.upgradeInbound(maConn, {
       signal: this.shutdownController.signal
     })
       .then(() => {
         this.log('inbound connection upgraded %s', maConn.remoteAddr)
 
-        socket.once('close', () => {
-          this.sockets.delete(socket)
-
-          if (
-            this.context.closeServerOnMaxConnections != null &&
-            this.sockets.size < this.context.closeServerOnMaxConnections.listenBelow
-          ) {
-            // The most likely case of error is if the port taken by this
-            // application is bound by another process during the time the
-            // server if closed. In that case there's not much we can do.
-            // resume() will be called again every time a connection is
-            // dropped, which acts as an eventual retry mechanism.
-            // onListenError allows the consumer act on this.
-            this.resume().catch(err => {
-              this.log.error('error attempting to listen server once connection count under limit - %e', err)
-              this.context.closeServerOnMaxConnections?.onListenError?.(err as Error)
-            })
-          }
-        })
-
         if (
           this.context.closeServerOnMaxConnections != null &&
           this.sockets.size >= this.context.closeServerOnMaxConnections.closeAbove
@@ -217,7 +206,6 @@ export class TCPListener extends TypedEventEmitter<ListenerEvents> implements Li
       .catch(async err => {
         this.log.error('inbound connection upgrade failed - %e', err)
         this.metrics.errors?.increment({ [`${this.addr} inbound_upgrade`]: true })
-        this.sockets.delete(socket)
         maConn.abort(err)
       })
   }
@@ -281,9 +269,14 @@ export class TCPListener extends TypedEventEmitter<ListenerEvents> implements Li
     // synchronously close any open connections - should be done after closing
     // the server socket in case new sockets are opened during the shutdown
     this.sockets.forEach(socket => {
-      if (socket.readable) {
-        events.push(pEvent(socket, 'close', options))
-        socket.destroy()
+         if (socket.closed) {
+             return;
+             }
+         events.push(pEvent(socket, 'close', {
+             ...options,
+             rejectionEvents: []
+         }));
+         socket.destroy();
       }
     })
 
diff --git a/src/tcp.ts b/src/tcp.ts
index 156347579d379786f31889b3e5a4b7aada996502..867fc330386b5aae4b5fe36f118173a80618f91a 100644
--- a/src/tcp.ts
+++ b/src/tcp.ts
@@ -28,8 +28,9 @@
  */
 
 import net from 'net'
-import { AbortError, TimeoutError, serviceCapabilities, transportSymbol } from '@libp2p/interface'
+import { AbortError, NotStartedError, TimeoutError, serviceCapabilities, transportSymbol } from '@libp2p/interface'
 import { TCP as TCPMatcher } from '@multiformats/multiaddr-matcher'
+import { setMaxListeners } from 'main-event'
 import { CustomProgressEvent } from 'progress-events'
 import { TCPListener } from './listener.js'
 import { toMultiaddrConnection } from './socket-to-conn.js'
@@ -44,11 +45,17 @@ export class TCP implements Transport<TCPDialEvents> {
   private readonly metrics?: TCPMetrics
   private readonly components: TCPComponents
   private readonly log: Logger
+  private readonly sockets: Set<Socket>
+  private shutdownController: AbortController
+
 
   constructor (components: TCPComponents, options: TCPOptions = {}) {
     this.log = components.logger.forComponent('libp2p:tcp')
     this.opts = options
     this.components = components
+    this.sockets = new Set()
+    this.shutdownController = new AbortController()
+    setMaxListeners(Infinity, this.shutdownController.signal)
 
     if (components.metrics != null) {
       this.metrics = {
@@ -72,7 +79,24 @@ export class TCP implements Transport<TCPDialEvents> {
     '@libp2p/transport'
   ]
 
+  start (): void {
+    this.shutdownController = new AbortController()
+    setMaxListeners(Infinity, this.shutdownController.signal)
+  }
+
+  async stop (): Promise<void> {
+    this.shutdownController.abort()
+
+    await Promise.all([...this.sockets].map(async socket => {
+      await this.closeSocket(socket)
+    }))
+  }
+
   async dial (ma: Multiaddr, options: TCPDialOptions): Promise<Connection> {
+    if (this.shutdownController.signal.aborted) {
+      throw new NotStartedError('TCP transport is stopped')
+    }
+
     options.keepAlive = options.keepAlive ?? true
     options.noDelay = options.noDelay ?? true
     options.allowHalfOpen = options.allowHalfOpen ?? false
@@ -123,6 +147,10 @@ export class TCP implements Transport<TCPDialEvents> {
 
       this.log('dialing %a with opts %o', ma, cOpts)
       rawSocket = net.connect(cOpts)
+      this.sockets.add(rawSocket)
+      rawSocket.once('close', () => {
+        this.sockets.delete(rawSocket)
+      })
 
       const onError = (err: Error): void => {
         this.log.error('dial to %a errored - %e', ma, err)
@@ -161,6 +189,7 @@ export class TCP implements Transport<TCPDialEvents> {
         if (options.signal != null) {
           options.signal.removeEventListener('abort', onAbort)
         }
+        this.shutdownController.signal.removeEventListener('abort', onAbort)
 
         if (err != null) {
           reject(err); return
@@ -174,13 +203,32 @@ export class TCP implements Transport<TCPDialEvents> {
       rawSocket.on('connect', onConnect)
 
       options.signal.addEventListener('abort', onAbort)
+      this.shutdownController.signal.addEventListener('abort', onAbort)
     })
-      .catch(err => {
-        rawSocket?.destroy()
+      .catch(async err => {
+        if (rawSocket != null) {
+          await this.closeSocket(rawSocket)
+        }
+
         throw err
       })
   }
 
+  private async closeSocket (socket: Socket): Promise<void> {
+    if (socket.closed) {
+      return
+    }
+
+    const closed = new Promise<void>(resolve => {
+      socket.once('close', () => {
+        resolve()
+      })
+    })
+
+    socket.destroy()
+    await closed
+  }
+
   /**
    * Creates a TCP listener. The provided `handler` function will be called
    * anytime a new incoming Connection has been successfully upgraded via
